add gnu Windows/iOS/Android builds and improve continuous delivery workflow

ygdrasil-io opened the pull request:
- Upgrade Upload and Download Actions to v4
- Simplify existing builds using a matrix strategy
- Add Windows x64 GNU build configuration
- Add iOS build configuration
- Add Android build configuration

ygdrasil-io commented:
Please don't merge yet. I'm currently working on adding mobile release (iOS and Android) and simplifying the workflow using Matrix.

ygdrasil-io committed:
Improve continuous delivery

- Upgrade Upload and Download Actions to v4
- Simplify existing builds using a matrix strategy
- Add Windows x64 GNU build configuration
- Add iOS build configuration
- Add Android build configuration

ygdrasil-io committed:
cleanup

ygdrasil-io commented:
For testing purposes, I created a fake release. If necessary, you can find the new builds here: 
https://github.com/ygdrasil-io/wgpu-native/releases/tag/v0.0.0.1

ygdrasil-io committed:
erratum - wrong macos archive name

ygdrasil-io commented:
https://github.com/ygdrasil-io/wgpu-native/releases/tag/v0.0.0.2 (There was a typo on Macos release)

The android version is working well :
![image](https://github.com/user-attachments/assets/67e666c6-5012-4485-8335-b45554772624)

almarklein commented:
This is great! I reviewed your changes, and it looks fine to me. Can you please also update the description at the top? It says "Each target (os + architecture) is build in a seperate job." but this is no longer the case.

Since this change is potentially impactful, I prefer to first merge #402 and do a release, then merge this pr and do another release.

almarklein commented:
We can combine this with  #410

almarklein commented:
Note: #411 will probably introduce merge conflicts.

ygdrasil-io committed:
Merge branch 'trunk' into feature/gnu-windows-static-library

almarklein commented on .github/workflows/cd.yml:
```diff
@@ -138,40 +60,97 @@ jobs:
+            target: i686-linux-android
+            setup_env: |
+              set -x
+              ANDROID_NDK_HOME=$ANDROID_HOME/ndk/25.2.9519653
```
Do you think we should use 27.0.12077973 as was done in #411?

ygdrasil-io replied:
It would be better to use the same version on both CI and CD pipelines. I'll try to find time this week to update this pull request and run thorough tests.

ygdrasil-io replied:
@almarklein No regressions were observed after updating.

![image](https://github.com/user-attachments/assets/eb2097ad-7431-44a2-8f20-6f57197780d9)

ygdrasil-io committed:
Merge branch 'gfx-rs:trunk' into feature/gnu-windows-static-library

ygdrasil-io committed:
Update the documentation at the top to accurately reflect the current jobs configuration.

ygdrasil-io committed:
Upgrade Android NDK to version 27.0.12077973 

Updated the Android NDK version from 25.2.9519653 to 27.0.12077973 across all target architectures in the GitHub Actions CI/CD pipeline. This change ensures compatibility with the latest NDK features and improvements. The `LIBCLANG_PATH` environment variable was also removed as it was no longer being utilized.

ygdrasil-io commented:
@almarklein If you have the permissions, could you please retry the failing job? The cloning of the Git repository https://github.com/glfw/glfw.git seems to have failed for some reason.

almarklein commented:
Awesome work, thanks! 🚀

almarklein commented:
Out of curiosity, do you know why the linux-aarch64 build takes abour half an hour, while all other builds take less than 4 minutes (most less than 1 min)? Would it be due to emulating arm64 on x86?

ygdrasil-io commented:
I think you are right, on my Mac Studio I see the same behaviour when I'm running a x86 VM, while arm VM are almost as quick of the host OS.

May be we could try cross compilation over virtualisation ?

almarklein commented:
It's not the end of the world, I was mostly curious because that one build stands out.
